File: TestObjects/CSCompiler/Method.cs
using System;
using System.Collections.Generic;
using System.Linq.Expressions;
using static ExpressionTreeTestObjects.Functions;
using static ExpressionTreeTestObjects.Categories;
using System.Linq;

namespace ExpressionTreeTestObjects {
    public static class Dummy {
        public static void DummyMethod() { }
        public static List<T> ExtensionMethod1Argument<T>(this List<T> lst, int n) => lst;
        public static List<T> ExtensionMethod2Arguments<T>(this List<T> lst, int n1, int n2) => lst;
        public static IEnumerable<T> ExtensionMethod1ArgumentEnumerable<T>(this IEnumerable<T> lst, int n) => lst;
        public static IEnumerable<T> ExtensionMethod2ArgumentsEnumerable<T>(this IEnumerable<T> lst, int n1, int n2) => lst;
        public static void DummyMethodWithGenerics<T>() { }
        public static void DummyMethodWithRef(ref int i) { }
        public static void DummyMethodWithOut(out int i) {
            i = 5;
        }
    }

    partial class CSCompiler {

        [TestObject(Method)]
        internal static readonly Expression InstanceMethod0Arguments = IIFE(() => {
            var s = "";
            return Expr(() => s.ToString());
        });

        [TestObject(Method)]
        internal static readonly Expression StaticMethod0Arguments = Expr(() => Dummy.DummyMethod());

        [TestObject(Method)]
        internal static readonly Expression ExtensionMethod0Arguments = IIFE(() => {
            var lst = new List<string>();
            return Expr(() => lst.Count());
        });

        [TestObject(Method)]
        internal static readonly Expression InstanceMethod1Argument = IIFE(() => {
            var s = "";
            return Expr(() => s.CompareTo(""));
        });

        [TestObject(Method)]
        internal static readonly Expression StaticMethod1Argument = Expr(() => string.Intern(""));

        [TestObject(Method)]
        internal static readonly Expression ExtensionMethod1Argument = IIFE(() => {
            var lst = new List<string>();
            return Expr(() => lst.ExtensionMethod1Argument(1));
        });

        [TestObject(Method)]
        internal static readonly Expression InstanceMethod2Arguments = IIFE(() => {
            var s = "abcde";
            return Expr(() => s.IndexOf('a', 2));
        });

        [TestObject(Method)]
        internal static readonly Expression StaticMethod2Arguments = Expr(() => string.Join(",", new[] { "a", "b" }));

        [TestObject(Method)]
        internal static readonly Expression ExtensionMethod2Arguments = IIFE(() => {
            var lst = new List<string>();
            return Expr(() => lst.ExtensionMethod2Arguments(5, 17));
        });

        [TestObject(Method)]
        internal static readonly Expression StringConcat = Expr((string s1, string s2) => string.Concat(s1, s2));

        [TestObject(Method)]
        internal static readonly Expression MathPow = Expr((double x, double y) => Math.Pow(x, y));

        [TestObject(Method)]
        internal static readonly Expression RequiredGenericParameters = Expr(() => Dummy.DummyMethodWithGenerics<string>());
    }
}

File: TestObjects.VB/VBCompiler/Method.vb
Imports ExpressionTreeTestObjects.Dummy

Partial Module VBCompiler
    <TestObject(Method)>
    Friend InstanceMethod0Arguments As Expression = IIFE(Function()
                                                             Dim s = ""
                                                             Return Expr(Function() s.ToString())
                                                         End Function)

    <TestObject(Method)>
    Friend StaticMethod0Arguments As Expression = Expr(Sub() DummyMethod())

    <TestObject(Method)>
    Friend ExtensionMethod0Arguments As Expression = IIFE(Function()
                                                              Dim lst = New List(Of String)()
                                                              Return Expr(Function() lst.Count)
                                                          End Function)

    <TestObject(Method)>
    Friend ExtensionMethod0ArgumentsWithoutConversion As Expression = IIFE(Function()
                                                                               Dim lst As IEnumerable(Of String) = New List(Of String)()
                                                                               Return Expr(Function() lst.Count)
                                                                           End Function)

    <TestObject(Method)>
    Friend InstanceMethod1Argument As Expression = IIFE(Function()
                                                            Dim s = ""
                                                            Return Expr(Function() s.CompareTo(""))
                                                        End Function)

    <TestObject(Method)>
    Friend StaticMethod1Argument As Expression = Expr(Function() String.Intern(""))

    <TestObject(Method)>
    Friend ExtensionMethod1Argument As Expression = IIFE(Function()
                                                             Dim lst = New List(Of String)()
                                                             Return Expr(Function() lst.ExtensionMethod1Argument(1))
                                                         End Function)

    <TestObject(Method)>
    Friend ExtensionMethod1ArgumentWithConversion As Expression = IIFE(Function()
                                                                           Dim lst = New List(Of String)()
                                                                           Return Expr(Function() lst.ExtensionMethod1ArgumentEnumerable(1))
                                                                       End Function)

    <TestObject(Method)>
    Friend ExtensionMethod1ArgumentWithoutConversion As Expression = IIFE(Function()
                                                                              Dim lst As IEnumerable(Of String) = New List(Of String)()
                                                                              Return Expr(Function() lst.ExtensionMethod1ArgumentEnumerable(1))
                                                                          End Function)

    <TestObject(Method)>
    Friend InstanceMethod2Arguments As Expression = IIFE(Function()
                                                             Dim s = "abcde"
                                                             Return Expr(Function() s.IndexOf("a"c, 2))
                                                         End Function)

    <TestObject(Method)>
    Friend StaticMethod2Arguments As Expression = IIFE(Function()
                                                           Dim arr = New Char() {"a"c, "b"c}
                                                           Return Expr(Function() String.Join(",", arr))
                                                       End Function)

    <TestObject(Method)>
    Friend StaticMethod2ArgumentsWithoutConversion As Expression = IIFE(Function()
                                                                            Dim arr As IEnumerable(Of Char) = New Char() {"a"c, "b"c}
                                                                            Return Expr(Function() String.Join(",", arr))
                                                                        End Function)

    <TestObject(Method)>
    Friend ExtensionMethod2Arguments As Expression = IIFE(Function()
                                                              Dim lst = New List(Of String)()
                                                              Return Expr(Function() lst.ExtensionMethod2Arguments(5, 17))
                                                          End Function)

    <TestObject(Method)>
    Friend ExtensionMethod2ArgumentsWithConversion As Expression = IIFE(Function()
                                                                            Dim lst = New List(Of String)()
                                                                            Return Expr(Function() lst.ExtensionMethod2ArgumentsEnumerable(5, 6))
                                                                        End Function)

    <TestObject(Method)>
    Friend ExtensionMethod2ArgumentsWithoutConversion As Expression = IIFE(Function()
                                                                               Dim lst As IEnumerable(Of String) = New List(Of String)()
                                                                               Return Expr(Function() lst.ExtensionMethod2ArgumentsEnumerable(5, 6))
                                                                           End Function)

    <TestObject(Method)>
    Friend StringConcat As Expression = Expr(Function(s1 As String, s2 As String) String.Concat(s1, s2))

    <TestObject(Method)>
    Friend StringConcatOperator As Expression = Expr(Function(s1 As String, s2 As String) s1 + s2)

    <TestObject(Method)>
    Friend StringConcatOperatorParamArray As Expression = Expr(Function(s1 As String, s2 As String) s1 + s2 + s1 + s2 + s1 + s2)

#If NETSTANDARD2_0 Then
#Else
    <TestObject(Method)>
    Friend LikeOperatorStrings As Expression = Expr(Function(s1 As String, s2 As String) s1 Like s2)

    <TestObject(Method)>
    Friend LikeOperatorObjects As Expression = Expr(Function(o1 As Object, o2 As Object) o1 Like o2)
#End If
End Module
